File: Samples/TextExtractTest/PHP/TextExtractTest.php
<?php
//---------------------------------------------------------------------------------------
// Copyright (c) 2001-2021 by PDFTron Systems Inc. All Rights Reserved.
// Consult LICENSE.txt regarding license information.
//---------------------------------------------------------------------------------------
include("../../../PDFNetC/Lib/PDFNetPHP.php");

// Relative path to the folder containing the test files.
$input_path = getcwd()."/../../TestFiles/newsletter.pdf";

//---------------------------------------------------------------------------------------
// This sample illustrates the basic text extraction capabilities of PDFNet.
//---------------------------------------------------------------------------------------

// A utility method used to dump all text content in the browser.
function DumpAllText($reader) 
{
	while (($element = $reader->Next()) != NULL)
	{
		switch ($element->GetType()) 
		{
		case Element::e_text_begin: 
			echo nl2br("\n--> Text Block Begin\n");
			break;
		case Element::e_text_end:
			echo nl2br("\n--> Text Block End\n");
			break;
		case Element::e_text:
			{
				$bbox = $element->GetBBox();
				echo nl2br("\n--> BBox: ".$bbox->x1.", "
							.$bbox->y1.", " 
							.$bbox->x2.", " 
							.$bbox->y2."\n");

				$arr = $element->GetTextString();
				echo nl2br($arr."\n");
			}
			break;
		case Element::e_text_new_line:
			echo nl2br("\n--> New Line\n");
			break;
		case Element::e_form:				// Process form XObjects
			$reader->FormBegin(); 
            		DumpAllText(reader);
			$reader->End(); 
			break; 
		}
	}
}

// A helper method for ReadTextFromRect
function RectTextSearch($reader, $pos) 
{		
	$srch_str = "";	
	while (($element = $reader->Next()) != null)
	{
		switch ($element->GetType()) 
		{
		case Element::e_text:
			{
				$bbox = $element->GetBBox();
				if($bbox->IntersectRect($bbox, $pos)) 
				{
					$arr = $element->GetTextString();
					$srch_str .= $arr;
					$srch_str .= nl2br("\n");
				}
				break;
			}
		case Element::e_text_new_line:
			{
				break;
			}
		case Element::e_form: // Process form XObjects
			{
				$reader->FormBegin(); 
				$srch_str .= RectTextSearch($reader, $pos);
				$reader->End(); 
				break; 
			}
		}
	}
	return $srch_str;
}

// A utility method used to extract all text content from
// a given selection rectangle. The rectangle coordinates are
// expressed in PDF user/page coordinate system.
function ReadTextFromRect($page, $pos, $reader)
{
	$reader->Begin($page);
	$str = RectTextSearch($reader, $pos);
	$reader->End();
	return $str;
}

function PrintStyle($style)
{
	$text_color = $style->GetColor();
	$tmp = sprintf("%02X%02X%02X;", $text_color[0], $text_color[1], $text_color[2]);
	echo " style=\"font-family:".$style->GetFontName()."; "
		."font-size:".$style->GetFontSize().";" 
		.($style->IsSerif() ? " sans-serif; " : " ")
		."color:#".$tmp."\"";
}

function IsStyleEqual($style1, $style2)
{
	if($style1->GetFontName() == $style2->GetFontName() && 
		$style1->GetFontSize() == $style1->GetFontSize() && 
		!($style1->IsSerif() xor $style1->IsSerif()) &&
		$style1->GetColor() == $style2->GetColor() ) {
		return true;
	}
	return false; 
}
//---------------------------------------------------------------------------------------

	PDFNet::Initialize();
	PDFNet::GetSystemFontList();    // Wait for fonts to be loaded if they haven't already. This is done because PHP can run into errors when shutting down if font loading is still in progress.

	$example1_basic     = false;
	$example2_xml       = false;
	$example3_wordlist  = false;
	$example4_advanced  = true;
	$example5_low_level = false;

	// Sample code showing how to use high-level text extraction APIs.
	
	$doc = new PDFDoc($input_path);
	$doc->InitSecurityHandler();

	$page = $doc->GetPage(1);
	if (!$page){
		echo nl2br("Page not found.\n");
		return;
	}

	$txt = new TextExtractor();
	$txt->Begin($page); // Read the page.
	// Other options you may want to consider...
	// txt.Begin(*itr, 0, TextExtractor::e_no_dup_remove);
	// txt.Begin(*itr, 0, TextExtractor::e_remove_hidden_text);

	// Example 1. Get all text on the page in a single string.
	// Words will be separated with space or new line characters.
	if ($example1_basic) 
	{
		// Get the word count.
		echo "Word Count: ".$txt->GetWordCount()."\n";

		$text = $txt->GetAsText();
		echo nl2br("\n\n- GetAsText --------------------------\n".$text."\n");
		echo nl2br("-----------------------------------------------------------\n");
	}

	// Example 2. Get XML logical structure for the page.
	if ($example2_xml) 
	{
		$text = $txt->GetAsXML(TextExtractor::e_words_as_elements | TextExtractor::e_output_bbox | TextExtractor::e_output_style_info);
		echo nl2br("\n\n- GetAsXML  --------------------------\n".$text."\n");
		echo nl2br("-----------------------------------------------------------\n");
	}

	// Example 3. Extract words one by one.
	if ($example3_wordlist) 
	{
		for ($line = $txt->GetFirstLine(); $line->IsValid(); $line=$line->GetNextLine())	{
			for ($word=$line->GetFirstWord(); $word->IsValid(); $word=$word->GetNextWord()) {
				echo nl2br($word->GetString()."\n");
			}
		}
		echo nl2br("-----------------------------------------------------------\n");
	}

	// Example 4. A more advanced text extraction example. 
	// The output is XML structure containing paragraphs, lines, words, 
	// as well as style and positioning information.
	if ($example4_advanced) 
	{
		$cur_flow_id=-1;
		$cur_para_id=-1;

		echo nl2br("<PDFText>\n");
		// For each line on the page...
		for ($line=$txt->GetFirstLine(); $line->IsValid(); $line=$line->GetNextLine())
		{
			if ($line->GetNumWords() == 0) continue;
			
			if ($cur_flow_id != $line->GetFlowID()) {
				if ($cur_flow_id != -1) {
					if ($cur_para_id != -1) {
						$cur_para_id = -1;
						echo nl2br("</Para>\n");
					}
					echo nl2br("</Flow>\n");
				}
				$cur_flow_id = $line->GetFlowID();
				echo nl2br("<Flow id=\"".$cur_flow_id."\">\n");
			}

			if ($cur_para_id != $line->GetParagraphID()) {
				if ($cur_para_id != -1)
					echo nl2br("</Para>\n");
				$cur_para_id = $line->GetParagraphID();
				echo nl2br("<Para id=\"".$cur_para_id."\">\n");
			}	

			$bbox1 = $line->GetBBox();
			$line_style = $line->GetStyle();
			printf("<Line box=\"%.2f, %.2f, %.2f, %.2f\"", $bbox1->x1, $bbox1->y1, $bbox1->x2, $bbox1->y2);
			PrintStyle($line_style);
			echo  " cur_num=\"".$line->GetCurrentNum()."\"";
			echo nl2br(">\n");

			// For each word in the line...
			for ($word=$line->GetFirstWord(); $word->IsValid(); $word=$word->GetNextWord())
			{
				// Output the bounding box for the word.
				$bbox2 = $word->GetBBox();
				printf("<Word box=\"%.2f, %.2f, %.2f, %.2f\"", $bbox2->x1, $bbox2->y1, $bbox2->x2, $bbox2->y2);
				echo " cur_num=\"" .$word->GetCurrentNum()."\"";
				$sz = $word->GetStringLen();
				if ($sz == 0) continue;

				// If the word style is different from the parent style, output the new style.
				$s = $word->GetStyle();
				if(!$s->IsEqual($line_style)){
					PrintStyle($s);
				}
				
				echo ">".$word->GetString();
				echo nl2br("</Word>\n");
			}
			echo nl2br("</Line>\n");
		}

		if ($cur_flow_id != -1) {
			if ($cur_para_id != -1) {
				$cur_para_id = -1;
				echo nl2br("</Para>\n");
			}
			echo nl2br("</Flow>\n");


		}
		echo nl2br("</PDFText>\n");

		$txt->Destroy();
		$doc->Close();

	}

	if($example5_low_level)
	{
		$doc = new PDFDoc($input_path);
		$doc->InitSecurityHandler();

		// Example 1. Extract all text content from the document

		$reader = new ElementReader();

		//  Read every page
		for ($itr=$doc->GetPageIterator(); $itr->HasNext(); $itr->Next()) 
		{
			$reader->Begin($itr->Current());
			DumpAllText($reader);
			$reader->End();
		}

		// Example 2. Extract text content based on the 
		// selection rectangle.
		echo nl2br("\n----------------------------------------------------");
		echo nl2br("\nExtract text based on the selection rectangle.");
		echo nl2br("\n----------------------------------------------------\n");

		$first_page = $doc->GetPage(1);
		$s1 = ReadTextFromRect($first_page, new Rect(27.0, 392.0, 563.0, 534.0), $reader);
		echo nl2br("\nField 1: ".$s1);

		$s1 = ReadTextFromRect($first_page, new Rect(28.0, 551.0, 106.0, 623.0), $reader);
		echo nl2br("\nField 2: ".$s1);

		$s1 = ReadTextFromRect($first_page, new Rect(208.0, 550.0, 387.0, 621.0), $reader);
		echo nl2br("\nField 3: ".$s1);

		// ... 
		$doc->Close();
		echo nl2br("Done.\n");
	}
?>

File: Samples/TextExtractTest/GO/TextExtract_test.go
//---------------------------------------------------------------------------------------
// Copyright (c) 2001-2023 by Apryse Software Inc. All Rights Reserved.
// Consult LICENSE.txt regarding license information.
//---------------------------------------------------------------------------------------

package main
import (
    "fmt"
    "testing"
    "strconv"
    "os"
    "flag"
    . "github.com/pdftron/pdftron-go/v2"
)

var licenseKey string
var modulePath string

func init() {
    flag.StringVar(&licenseKey, "license", "", "License key for Apryse SDK")
    flag.StringVar(&modulePath, "modulePath", "", "Module path for Apryse SDK")
}

func PrintStyle (style Style){
    sansSerifStr := ""
    if style.IsSerif(){
        sansSerifStr = " sans-serif;"
	}
    rgb := style.GetColor()
    rgbHex := fmt.Sprintf("%02X%02X%02X;", rgb.Get(0), rgb.Get(1), rgb.Get(2))
    fontStr := fmt.Sprintf("%g", style.GetFontSize())
    os.Stdout.Write([]byte(" style=\"font-family:" + style.GetFontName() + "; font-size:" + fontStr + ";" + sansSerifStr + " color:#" + rgbHex + "\""))
}

func DumpAllText (reader ElementReader){
    element := reader.Next()

    for element.GetMp_elem().Swigcptr() != 0{
        etype := element.GetType()
        if etype == ElementE_text_begin{
            fmt.Println("Text Block Begin")
        }else if etype == ElementE_text_end{
            fmt.Println("Text Block End")
        }else if etype == ElementE_text{
            bbox := element.GetBBox()
            fmt.Println("BBox: " + fmt.Sprintf("%f", bbox.GetX1()) + ", " + fmt.Sprintf("%f", bbox.GetY1()) + ", " +
						fmt.Sprintf("%f", bbox.GetX2()) + ", " + fmt.Sprintf("%f", bbox.GetY2()))
            textString := element.GetTextString()
            fmt.Println(textString)
        }else if etype == ElementE_text_new_line{
            fmt.Println("New Line")
        }else if etype == ElementE_form{
            reader.FormBegin()
            DumpAllText(reader)
            reader.End()
		}
        element = reader.Next()
	}
}

// A utility method used to extract all text content from
// a given selection rectangle. The recnagle coordinates are
// expressed in PDF user/page coordinate system.
func ReadTextFromRect (page Page, pos Rect, reader ElementReader) string{
    reader.Begin(page)
    srchStr := RectTextSearch(reader, pos)
    reader.End()
    return srchStr
}
//A helper method for ReadTextFromRect
func RectTextSearch (reader ElementReader, pos Rect) string{
    element := reader.Next()
    srchStr2 := ""
    for element.GetMp_elem().Swigcptr() != 0{
        etype := element.GetType()
        if etype == ElementE_text{
            bbox := element.GetBBox()
            if (bbox.IntersectRect(bbox, pos)){
                arr := element.GetTextString()
                srchStr2 += arr
                srchStr2 += "\n"
			}
        }else if etype == ElementE_text_new_line{
            //handle text new line here
        }else if etype == ElementE_form{
            reader.FormBegin()
            srchStr2 += RectTextSearch(reader, pos)
            fmt.Println(srchStr2)
            reader.End()
		}
        element = reader.Next()
	}
    return srchStr2
}

func TestTextExtract(t *testing.T){
    PDFNetInitialize(licenseKey)
    
    // Relative path to the folder containing test files.
    inputPath :=  "../TestFiles/newsletter.pdf"
    example1Basic := false
    example2Xml := false
    example3Wordlist := false
    example4Advanced := true
    example5LowLevel := false
   
    // Sample code showing how to use high-level text extraction APIs.
    doc := NewPDFDoc(inputPath)
    doc.InitSecurityHandler()
    
    page := doc.GetPage(1)
    if page == nil{
        fmt.Println("page no found")
    }    
    txt := NewTextExtractor()
    txt.Begin(page) // Read the page
    
    // Example 1. Get all text on the page in a single string.
    // Words will be separated witht space or new line characters.
    if example1Basic{
        fmt.Println("Word count: " + strconv.Itoa(txt.GetWordCount()))
        txtAsText := txt.GetAsText()
        fmt.Println("- GetAsText --------------------------" + txtAsText)
        fmt.Println("-----------------------------------------------------------")
	}
    // Example 2. Get XML logical structure for the page.
    if example2Xml{
        text := txt.GetAsXML(TextExtractorE_words_as_elements | 
                            TextExtractorE_output_bbox | 
                            TextExtractorE_output_style_info)       
        fmt.Println("- GetAsXML  --------------------------" + text)
        fmt.Println("-----------------------------------------------------------")
    }
    // Example 3. Extract words one by one.
    if example3Wordlist{
        word := NewWord()
        line := txt.GetFirstLine()
        for line.IsValid(){
            word = line.GetFirstWord()
            for word.IsValid(){
                wordString := word.GetString()
                fmt.Println(wordString)
                word = word.GetNextWord()
			}
            line = line.GetNextLine()
		}
        fmt.Println("-----------------------------------------------------------")
	}
    // Example 4. A more advanced text extraction example. 
    // The output is XML structure containing paragraphs, lines, words, 
    // as well as style and positioning information.
    if example4Advanced{
        bbox := NewRect()
        curFlowId := -1
        curParaId := -1
        
        fmt.Println("<PDFText>")
        // For each line on the page...
        line := txt.GetFirstLine()
        for line.IsValid(){
            if line.GetNumWords() == 0{
                line = line.GetNextLine()			
                continue
			}
            word := line.GetFirstWord()
            if curFlowId != line.GetFlowID(){
                if curFlowId != -1{
                    if curParaId != -1{
                        curParaId = -1
                        fmt.Println("</Para>")
					}
                    fmt.Println("</Flow>")
				}
                curFlowId = line.GetFlowID()
                fmt.Println("<Flow id=\"" + strconv.Itoa(curFlowId) +"\">")
            }        
            if curParaId != line.GetParagraphID(){
                if curParaId != -1{
                    fmt.Println("</Para>")
				}
                curParaId= line.GetParagraphID()
                fmt.Println("<Para id=\"" +strconv.Itoa(curParaId)+ "\">")
            }    
            bbox = line.GetBBox()
            lineStyle := line.GetStyle()
            os.Stdout.Write([]byte(fmt.Sprintf("<Line box=\"%.2f, %.2f, %.2f, %.2f\"", bbox.GetX1(), bbox.GetY1(), bbox.GetX2(), bbox.GetY2())))
            PrintStyle (lineStyle)
            os.Stdout.Write([]byte(" cur_num=\"" + strconv.Itoa(line.GetCurrentNum()) + "\"" + ">\n"))
            
            // For each word in the line...
            word = line.GetFirstWord()
            for word.IsValid(){
                // Output the bounding box for the word
                bbox = word.GetBBox()
				os.Stdout.Write([]byte(fmt.Sprintf("<Word box=\"%.2f, %.2f, %.2f, %.2f\"", bbox.GetX1(), bbox.GetY1(), bbox.GetX2(), bbox.GetY2())))
                os.Stdout.Write([]byte(" cur_num=\"" + strconv.Itoa(word.GetCurrentNum()) + "\""));
                sz := word.GetStringLen()
                if sz == 0{
                    word = word.GetNextWord()				
                    continue
				}
                // If the word style is different from the parent style, output the new style.
                s := word.GetStyle()
                if !s.IsEqual(lineStyle){
                    PrintStyle (s)
				}
                wordString := word.GetString()
                os.Stdout.Write([]byte(">" + wordString + "</Word>\n"))
                word = word.GetNextWord()
			}
            os.Stdout.Write([]byte("</Line>\n"))               
            line = line.GetNextLine()
        }    
        if curFlowId != -1{
            if curParaId != -1{
                curParaId = -1
                os.Stdout.Write([]byte("</Para>\n"))
			}
            os.Stdout.Write([]byte("</Flow>\n"))
        }
        txt.Destroy()
        doc.Close()            
        fmt.Println("</PDFText>")
    }
    // Sample code showing how to use low-level text extraction APIs.
    if example5LowLevel{
        doc = NewPDFDoc(inputPath)
        doc.InitSecurityHandler()

        // Example 1. Extract all text content from the document
        
        reader := NewElementReader()
        itr := doc.GetPageIterator()
        for itr.HasNext(){
            reader.Begin(itr.Current())
            DumpAllText(reader)
            reader.End()
            itr.Next()
        }
		
        // Example 2. Extract text content based on the 
        // selection rectangle.
        
        fmt.Println("----------------------------------------------------")
        fmt.Println("Extract text based on the selection rectangle.")
        fmt.Println("----------------------------------------------------")
        
        itr = doc.GetPageIterator()
        firstPage := itr.Current()
        s1 := ReadTextFromRect(firstPage, NewRect(27.0, 392.0, 563.0, 534.0), reader)
        fmt.Println("Field 1: " + s1)

        s1 = ReadTextFromRect(firstPage, NewRect(28.0, 551.0, 106.0, 623.0), reader);
        fmt.Println("Field 2: " + s1)

        s1 = ReadTextFromRect(firstPage, NewRect(208.0, 550.0, 387.0, 621.0), reader);
        fmt.Println("Field 3: " + s1)
        
        doc.Close()
        PDFNetTerminate()
        fmt.Println("Done.")
	}
}
